### PR TITLE
[X86] Finally handle target of __builtin_ia32_cmp[p|s][s|d] from avx into sse/sse2/avx

### PR DESCRIPTION
**Pull request by FreddyLeaf**
This patch relands #67410 and fixes the cmpfail below:
#include <immintrin.h>
__attribute__((target("avx"))) void test(__m128 a, __m128 b) {
  _mm_cmp_ps(a, b, 14);
}

According to Intel SDM, SSE/SSE2 instructions cmp[p|s][s|d] are supported when imm8 is in range of [0, 7]

### CODE REVIEW

File: clang/test/CodeGen/target-features-error-2.c [hunk 1]
```diff
@@ -15,7 +15,7 @@ int baz(__m256i a) {
 
 #if NEED_AVX_2
 __m128 need_avx(__m128 a, __m128 b) {
-  return _mm_cmp_ps(a, b, 0); // expected-error {{'__builtin_ia32_cmpps' needs target feature avx}}
+  return _mm_cmp_ps(a, b, 8); // expected-error {{'__builtin_ia32_cmpps' needs target feature avx}}
```
Comment (FreddyLeaf) [inline clang/test/CodeGen/target-features-error-2.c:18]: @phoebewang @RKSimon @KanRobert 
This change shows the behavior of compiling these intrinsics with parameter of [8,31] after this PR.

**Comment by FreddyLeaf** [comment]
The difference with #67410 is here: 54c78b5ad694946186ba115576a6eea9c8992349

File: clang/include/clang/Basic/BuiltinsX86.def [hunk 2]
```diff
@@ -462,12 +462,12 @@ TARGET_BUILTIN(__builtin_ia32_blendvps256, "V8fV8fV8fV8f", "ncV:256:", "avx")
...
-TARGET_BUILTIN(__builtin_ia32_cmppd, "V2dV2dV2dIc", "ncV:128:", "avx")
+TARGET_BUILTIN(__builtin_ia32_cmppd, "V2dV2dV2dIc", "ncV:128:", "avx|sse2")
 TARGET_BUILTIN(__builtin_ia32_cmppd256, "V4dV4dV4dIc", "ncV:256:", "avx")
-TARGET_BUILTIN(__builtin_ia32_cmpps, "V4fV4fV4fIc", "ncV:128:", "avx")
+TARGET_BUILTIN(__builtin_ia32_cmpps, "V4fV4fV4fIc", "ncV:128:", "avx|sse")
 TARGET_BUILTIN(__builtin_ia32_cmpps256, "V8fV8fV8fIc", "ncV:256:", "avx")
-TARGET_BUILTIN(__builtin_ia32_cmpsd, "V2dV2dV2dIc", "ncV:128:", "avx")
-TARGET_BUILTIN(__builtin_ia32_cmpss, "V4fV4fV4fIc", "ncV:128:", "avx")
+TARGET_BUILTIN(__builtin_ia32_cmpsd, "V2dV2dV2dIc", "ncV:128:", "avx|sse2")
+TARGET_BUILTIN(__builtin_ia32_cmpss, "V4fV4fV4fIc", "ncV:128:", "avx|sse")
```
Comment (RKSimon) [inline clang/include/clang/Basic/BuiltinsX86.def:470]: Why not move these up to the SSE/SSE2 implementations? And is the "avx|" component necessary anymore?

File: clang/lib/CodeGen/CodeGenFunction.cpp [hunk 3]
```diff
@@ -2613,6 +2614,24 @@ void CGBuilderInserter::InsertHelper(
 // called function.
 void CodeGenFunction::checkTargetFeatures(const CallExpr *E,
                                           const FunctionDecl *TargetDecl) {
+  // SemaCheking cannot handle below x86 builtins because they have different
```
Comment (RKSimon) [inline clang/lib/CodeGen/CodeGenFunction.cpp:2617]: Semachecking

File: clang/test/CodeGen/X86/sse-builtins.c [hunk 4]
```diff
@@ -813,3 +813,57 @@ __m128 test_mm_xor_ps(__m128 A, __m128 B) {
   // CHECK: xor <4 x i32>
   return _mm_xor_ps(A, B);
 }
+
+__m128 test_mm_cmp_ps_eq_oq(__m128 a, __m128 b) {
```
Comment (RKSimon) [inline clang/test/CodeGen/X86/sse-builtins.c:817]: Move these up (they should be approximately alpha sorted)

File: clang/test/CodeGen/X86/sse2-builtins.c [hunk 5]
```diff
@@ -1719,3 +1719,57 @@ __m128i test_mm_xor_si128(__m128i A, __m128i B) {
   // CHECK: xor <2 x i64> %{{.*}}, %{{.*}}
   return _mm_xor_si128(A, B);
 }
+
+__m128d test_mm_cmp_pd_eq_oq(__m128d a, __m128d b) {
```
Comment (RKSimon) [inline clang/test/CodeGen/X86/sse2-builtins.c:1723]: Sorting

**Review by RKSimon — COMMENTED** [review]
Should we update the existing SSE/SSE2 cmpeq/cmplt intinsics to use __builtin_ia32_cmp* instead of __builtin_ia32_cmpeq??/__builtin_ia32_cmplt?? etc?

File: clang/include/clang/Basic/BuiltinsX86.def (hunk 2 above)
Comment (KanRobert) [inline clang/include/clang/Basic/BuiltinsX86.def:470]: Why do we need "avx|"?

Move this to line 346?

**Comment by KanRobert** [comment]
My understanding is that you change the required feature of __builtin_ia32_cmp[p|s][s|d] from avx to sse for some condition codes. The title is not accurate.

File: clang/lib/CodeGen/CodeGenFunction.cpp [hunk 6]
```diff
@@ -2613,6 +2614,24 @@ void CGBuilderInserter::InsertHelper(
...
+  // parameter ranges with different TargetAttribute of caller.
```
Comment (KanRobert) [inline clang/lib/CodeGen/CodeGenFunction.cpp:2618]: Why? Any background. Please add it to the description.

**Comment by FreddyLeaf** [comment]
> Should we update the existing SSE/SSE2 cmpeq/cmplt intinsics to use __builtin_ia32_cmp* instead of __builtin_ia32_cmpeq??/__builtin_ia32_cmplt?? etc?

Good question! Didn't notice this issue before.

File: clang/include/clang/Basic/BuiltinsX86.def (hunk 2 above)
Comment (FreddyLeaf) [reply clang/include/clang/Basic/BuiltinsX86.def:470]: Good catch. No strong objections.

Comment (FreddyLeaf) [reply clang/include/clang/Basic/BuiltinsX86.def:470]: Good catch. No strong objections.

**Comment by FreddyLeaf** [comment]
> My understanding is that you change the required feature of __builtin_ia32_cmp[p|s][s|d] from avx to sse for some condition codes. The title is not accurate.

You are right. Updated.

File: clang/lib/CodeGen/CodeGenFunction.cpp (hunk 6 above)
Comment (FreddyLeaf) [reply clang/lib/CodeGen/CodeGenFunction.cpp:2618]: Semachecking doesn't have the information of the caller of the intrinsic. While here we require to know if caller has "avx" or not when the condition code is [8, 31].
Considering such case:
```
#include <immintrin.h>

__attribute__((target("sse2")))
__m128 test1(__m128 a, __m128 b) {
  return _mm_cmp_ps(a, b, 14); // shouldn't report error.
}

__attribute__((target("avx")))
__m128 test2(__m128 a, __m128 b) {
  return _mm_cmp_ps(a, b, 14); // should report error.
}

Comment (FreddyLeaf) [reply clang/lib/CodeGen/CodeGenFunction.cpp]: [d8979e2](https://github.com/llvm/llvm-project/pull/84136/commits/d8979e29e1932b42bb54424e1c6aa97ba9ba07e7)

Comment (FreddyLeaf) [reply clang/lib/CodeGen/CodeGenFunction.cpp]: [d8979e2](https://github.com/llvm/llvm-project/pull/84136/commits/d8979e29e1932b42bb54424e1c6aa97ba9ba07e7)

File: clang/include/clang/Basic/BuiltinsX86.def (hunk 2 above)
Comment (FreddyLeaf) [reply clang/include/clang/Basic/BuiltinsX86.def]: [d8979e2](https://github.com/llvm/llvm-project/pull/84136/commits/d8979e29e1932b42bb54424e1c6aa97ba9ba07e7)

File: clang/test/CodeGen/X86/sse2-builtins.c (hunk 5 above)
Comment (FreddyLeaf) [reply clang/test/CodeGen/X86/sse2-builtins.c]: [d8979e2](https://github.com/llvm/llvm-project/pull/84136/commits/d8979e29e1932b42bb54424e1c6aa97ba9ba07e7)

File: clang/test/CodeGen/X86/sse-builtins.c (hunk 4 above)
Comment (FreddyLeaf) [reply clang/test/CodeGen/X86/sse-builtins.c]: [d8979e2](https://github.com/llvm/llvm-project/pull/84136/commits/d8979e29e1932b42bb54424e1c6aa97ba9ba07e7)

File: clang/lib/CodeGen/CodeGenFunction.cpp (hunk 3 above)
Comment (FreddyLeaf) [reply clang/lib/CodeGen/CodeGenFunction.cpp]: [d8979e2](https://github.com/llvm/llvm-project/pull/84136/commits/d8979e29e1932b42bb54424e1c6aa97ba9ba07e7)

**Comment by FreddyLeaf** [comment]
> > Should we update the existing SSE/SSE2 cmpeq/cmplt intinsics to use __builtin_ia32_cmp* instead of __builtin_ia32_cmpeq??/__builtin_ia32_cmplt?? etc?
> 
> Good question! Didn't notice this issue before.

Seems doable. I'll prepare a following PR after this.

File: clang/lib/Headers/emmintrin.h [hunk 7]
```diff
@@ -4745,6 +4745,127 @@ static __inline__ __m128d __DEFAULT_FN_ATTRS _mm_castsi128_pd(__m128i __a) {
...
+/// \endcode
+///
+/// This intrinsic corresponds to the <c> (V)CMPPD </c> instruction.
+///
+/// \param a
+///    A 128-bit vector of [2 x double].
+/// \param b
+///    A 128-bit vector of [2 x double].
+/// \param c
+///    An immediate integer operand, with bits [4:0] specifying which comparison
```
Comment (RKSimon) [inline clang/lib/Headers/emmintrin.h:4769]: These need updating to indicate which are AVX-only

Comment (FreddyLeaf) [reply clang/lib/Headers/emmintrin.h:4769]: There's such note added.
```
(Note that without avx enabled, only bits [2:0] are supported)

File: clang/lib/Headers/avxintrin.h [hunk 8]
```diff
@@ -1573,15 +1573,6 @@ _mm256_blendv_ps(__m256 __a, __m256 __b, __m256 __c)
...
 
-/* Compare */
-#define _CMP_EQ_OQ    0x00 /* Equal (ordered, non-signaling)  */
-#define _CMP_LT_OS    0x01 /* Less-than (ordered, signaling)  */
-#define _CMP_LE_OS    0x02 /* Less-than-or-equal (ordered, signaling)  */
-#define _CMP_UNORD_Q  0x03 /* Unordered (non-signaling)  */
-#define _CMP_NEQ_UQ   0x04 /* Not-equal (unordered, non-signaling)  */
-#define _CMP_NLT_US   0x05 /* Not-less-than (unordered, signaling)  */
-#define _CMP_NLE_US   0x06 /* Not-less-than-or-equal (unordered, signaling)  */
-#define _CMP_ORD_Q    0x07 /* Ordered (non-signaling)   */
```
Comment (phoebewang) [inline clang/lib/Headers/avxintrin.h:1584]: Why removing them?

File: clang/lib/Headers/xmmintrin.h [hunk 9]
```diff
@@ -2940,6 +2940,134 @@ _mm_movemask_ps(__m128 __a)
...
+///    0x16: Not-less-than-or-equal (unordered, non-signaling) \n
+///    0x17: Ordered (signaling) \n
+///    0x18: Equal (unordered, signaling) \n
+///    0x19: Not-greater-than-or-equal (unordered, non-signaling) \n
+///    0x1A: Not-greater-than (unordered, non-signaling) \n
+///    0x1B: False (ordered, signaling) \n
+///    0x1C: Not-equal (ordered, signaling) \n
+///    0x1D: Greater-than-or-equal (ordered, non-signaling) \n
+///    0x1E: Greater-than (ordered, non-signaling) \n
+///    0x1F: True (unordered, signaling)
```
Comment (phoebewang) [inline clang/lib/Headers/xmmintrin.h:3008]: I think we either give a full define of the macros or remove these comments.

File: clang/test/CodeGen/X86/sse-builtins.c [hunk 10]
```diff
@@ -34,6 +34,60 @@ __m128 test_mm_andnot_ps(__m128 A, __m128 B) {
   return _mm_andnot_ps(A, B);
 }
 
+__m128 test_mm_cmp_ps_eq_oq(__m128 a, __m128 b) {
+  // CHECK-LABEL: test_mm_cmp_ps_eq_oq
+  // CHECK: fcmp oeq <4 x float> %{{.*}}, %{{.*}}
+  return _mm_cmp_ps(a, b, _CMP_EQ_OQ);
```
Comment (phoebewang) [inline clang/test/CodeGen/X86/sse-builtins.c:40]: I don't understand why moving them here. Shouldn't they emit error without avx?

File: clang/lib/Headers/xmmintrin.h (hunk 9 above)
Comment (RKSimon) [reply clang/lib/Headers/xmmintrin.h:3008]: No objections, but I'd prefer to see the new macros (and tests) and comment removal to be done as a followup patch

File: clang/test/CodeGen/X86/sse-builtins.c (hunk 10 above)
Comment (FreddyLeaf) [reply clang/test/CodeGen/X86/sse-builtins.c:40]: These condition codes are [0,7], so they won't throw errors. What about adding 8 more tests to avx-builtins.c with condition codes of [8,15]?

File: clang/lib/CodeGen/CodeGenFunction.cpp (hunk 6 above)
Comment (KanRobert) [reply clang/lib/CodeGen/CodeGenFunction.cpp]: This is the behavior. But why?

Comment (FreddyLeaf) [reply clang/lib/CodeGen/CodeGenFunction.cpp]: because SDM defined so?

Comment (KanRobert) [reply clang/lib/CodeGen/CodeGenFunction.cpp]: I see. They corresponds to different instructions.

File: clang/test/CodeGen/X86/sse-builtins.c (hunk 10 above)
Comment (FreddyLeaf) [reply clang/test/CodeGen/X86/sse-builtins.c:40]: Oh the remaining 24 are already there. So I only moved the first 8 from avx to sse/sse2. which is expected.

File: clang/lib/Headers/xmmintrin.h (hunk 9 above)
Comment (FreddyLeaf) [reply clang/lib/Headers/xmmintrin.h]: Refined in 51eda069c797c91fd3072f347e1dd4a20ab53702

File: clang/lib/Headers/emmintrin.h (hunk 7 above)
Comment (FreddyLeaf) [reply clang/lib/Headers/emmintrin.h:4769]: Refined in 51eda069c797c91fd3072f347e1dd4a20ab53702

File: clang/lib/CodeGen/CodeGenFunction.cpp (hunk 6 above)
Comment (FreddyLeaf) [reply clang/lib/CodeGen/CodeGenFunction.cpp]: Updated in description.

**Comment by KanRobert** [comment]
Still not understand. Shouldn't the target feature of the caller be passed to the intrinsic when checking the argument of the intrinsic?

**Comment by FreddyLeaf** [comment]
> Still not understand. Shouldn't the target feature of the caller be passed to the intrinsic when checking the argument of the intrinsic?
The target features passing from option can be passed to Semachecking. It's special for attribute_target, which is handled in `handleTargetAttr`, which is after Sema::CheckBuiltinFunctionCall.

**Comment by KanRobert** [comment]
> > Still not understand. Shouldn't the target feature of the caller be passed to the intrinsic when checking the argument of the intrinsic?
> 
> The target features passing from option can be passed to Semachecking. It's special for attribute_target, which is handled in `handleTargetAttr`, which is after Sema::CheckBuiltinFunctionCall.

Then why not move the call `handleTargetAttr` before `Sema::CheckBuiltinFunctionCall`?

**Comment by FreddyLeaf** [comment]
> > > Still not understand. Shouldn't the target feature of the caller be passed to the intrinsic when checking the argument of the intrinsic?
> > 
> > 
> > The target features passing from option can be passed to Semachecking. It's special for attribute_target, which is handled in `handleTargetAttr`, which is after Sema::CheckBuiltinFunctionCall.
> 
> Then why not move the call `handleTargetAttr` before `Sema::CheckBuiltinFunctionCall`?

Mainly because all Sema::CheckBuiltinFunctionCall before doesn't dependent on target feature, right? We don't have any other builtins have different range limit of same const parameter. But clang::CodegenFunction has the information of the whole function, which I think is a more reasonable place to do the check.

**Comment by FreddyLeaf** [comment]
ping for review

**Review by phoebewang — APPROVED** [review]
LGTM.

**Review by KanRobert — APPROVED** [review]
LGTM

**Review by RKSimon — APPROVED** [review]
LGTM - but @pogo59 needs a headup as it will affect some ongoing documentation cleanup work

File: clang/lib/Headers/emmintrin.h [hunk 11]
```diff
@@ -4745,6 +4745,77 @@ static __inline__ __m128d __DEFAULT_FN_ATTRS _mm_castsi128_pd(__m128i __a) {
   return (__m128d)__a;
 }
 
+/// Compares each of the corresponding double-precision values of two
+///    128-bit vectors of [2 x double], using the operation specified by the
+///    immediate integer operand.
+///
+///    Returns a [2 x double] vector consisting of two doubles corresponding to
+///    the two comparison results: zero if the comparison is false, and all 1's
+///    if the comparison is true.
```
Comment (pogo59) [inline clang/lib/Headers/emmintrin.h:4754]:
```suggestion
///    Each comparison yields 0x0 for false, 0xFFFFFFFFFFFFFFFF for true.
```
to be consistent with #83316 changes.

File: clang/lib/Headers/emmintrin.h [hunk 12]
```diff
@@ -4745,6 +4745,77 @@ static __inline__ __m128d __DEFAULT_FN_ATTRS _mm_castsi128_pd(__m128i __a) {
...
+#define _mm_cmp_pd(a, b, c)                                                    \
+  ((__m128d)__builtin_ia32_cmppd((__v2df)(__m128d)(a), (__v2df)(__m128d)(b),   \
+                                 (c)))
+
+/// Compares each of the corresponding scalar double-precision values of
+///    two 128-bit vectors of [2 x double], using the operation specified by the
+///    immediate integer operand.
+///
+///    If the result is true, all 64 bits of the destination vector are set;
+///    otherwise they are cleared.
```
Comment (pogo59) [inline clang/lib/Headers/emmintrin.h:4789]:
```suggestion
///    Each comparison yields 0x0 for false, 0xFFFFFFFFFFFFFFFF for true.
```

File: clang/lib/Headers/xmmintrin.h [hunk 13]
```diff
@@ -2940,6 +2940,84 @@ _mm_movemask_ps(__m128 __a)
...
+#define _CMP_NLE_US   0x06 /* Not-less-than-or-equal (unordered, signaling)  */
+#define _CMP_ORD_Q    0x07 /* Ordered (non-signaling)   */
+
+/// Compares each of the corresponding values of two 128-bit vectors of
+///    [4 x float], using the operation specified by the immediate integer
+///    operand.
+///
+///    Returns a [4 x float] vector consisting of four floats corresponding to
+///    the four comparison results: zero if the comparison is false, and all 1's
+///    if the comparison is true.
```
Comment (pogo59) [inline clang/lib/Headers/xmmintrin.h:2959]:
```suggestion
///    Each comparison yields 0x0 for false, 0xFFFFFFFF for true.
```

File: clang/lib/Headers/xmmintrin.h [hunk 14]
```diff
@@ -2940,6 +2940,84 @@ _mm_movemask_ps(__m128 __a)
...
+/// \returns A 128-bit vector of [4 x float] containing the comparison results.
+#define _mm_cmp_ps(a, b, c)                                                    \
+  ((__m128)__builtin_ia32_cmpps((__v4sf)(__m128)(a), (__v4sf)(__m128)(b), (c)))
+
+/// Compares each of the corresponding scalar values of two 128-bit
+///    vectors of [4 x float], using the operation specified by the immediate
+///    integer operand.
+///
+///    If the result is true, all 32 bits of the destination vector are set;
+///    otherwise they are cleared.
```
Comment (pogo59) [inline clang/lib/Headers/xmmintrin.h:2993]:
```suggestion
///    Each comparison yields 0x0 for false, 0xFFFFFFFF for true.
```

**Review by pogo59 — COMMENTED** [review]
Thanks for the heads-up @RKSimon I've made suggestions for consistency with existing practice in emmintrin.h which I've adopted in the latest revision to #83316.

I'm unsure how replacing a function declaration with an \fn directive will play with our documentation tooling. But we can figure that out later.

**Review by pogo59 — COMMENTED** [review]
The example for `\fn` in the [doxygen documentation](https://www.doxygen.nl/manual/commands.html#cmdfn) puts the directive at the top of the comment block, not at the end. Do we know whether putting it at the end will produce the expected result?

File: clang/lib/Headers/xmmintrin.h (hunk 14 above)
Comment (FreddyLeaf) [reply clang/lib/Headers/xmmintrin.h]: 71641346ebadf8d21a109813fcf66f78e0932955

File: clang/lib/Headers/xmmintrin.h (hunk 13 above)
Comment (FreddyLeaf) [reply clang/lib/Headers/xmmintrin.h]: 71641346ebadf8d21a109813fcf66f78e0932955

File: clang/lib/Headers/emmintrin.h (hunk 12 above)
Comment (FreddyLeaf) [reply clang/lib/Headers/emmintrin.h]: 71641346ebadf8d21a109813fcf66f78e0932955

File: clang/lib/Headers/emmintrin.h (hunk 11 above)
Comment (FreddyLeaf) [reply clang/lib/Headers/emmintrin.h]: 71641346ebadf8d21a109813fcf66f78e0932955

**Comment by FreddyLeaf** [comment]
> Thanks for the heads-up @RKSimon I've made suggestions for consistency with existing practice in emmintrin.h which I've adopted in the latest revision to #83316.
> 
> I'm unsure how replacing a function declaration with an \fn directive will play with our documentation tooling. But we can figure that out later.

I've committed your suggestions. FYI \fn usage was referring https://github.com/llvm/llvm-project/blob/c58c8278f98c189d149d5f062b8d4f56efcada90/clang/lib/Headers/avxvnniintrin.h#L32

**Comment by pogo59** [comment]
Thanks!

**Pull request merged**